File: finance-tracker-ui/src/app/core/budget.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Budget {
  id?: number;
  categoryId: number;
  amount: number;
  month: number;
  year: number;
  category?: {name: string; type: string};
}

@Injectable({
  providedIn: 'root',
})
export class BudgetService {
  private apiUrl = "https://za-finance-tracker.onrender.com/api/Budgets";
  constructor(private http: HttpClient) {}

  getAll(): Observable<Budget[]> {
    return this.http.get<Budget[]>(this.apiUrl);
  }

  create(budget: Budget): Observable<Budget> {
    return this.http.post<Budget>(this.apiUrl, budget);
  }
  
  update(id: number, budget: Budget): Observable<Budget> {
    return this.http.put<Budget>(`${this.apiUrl}/${id}`, budget);
  } 

  delete(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`, { responseType: 'text' });
  }
}

File: finance-tracker-api/Models/Budget.cs
namespace finance_tracker_api.Models
{
    public class Budget
    {
        public int Id {get; set;}
        public int UserId {get; set;}
        public int CategoryId {get; set;}
        public decimal Amount {get; set;}
        public int Month {get; set;} // 1-12
        public int Year {get; set;}
        
        public User? User {get; set;}
        public Category? Category {get; set;} = null!;
    }
}